File: client/src/keypool.ts
/**
 * FlockKeyPool — multi-key rotation with 429/Retry-After tracking.
 *
 * Absorbed from the standalone super-ralph nimProxy integration so the
 * flock-client SDK carries the same resilience without a proxy in front:
 * keys are deduplicated on registration, rotation is round-robin over
 * keys whose rate limit has expired, and a 429 parks the offending key
 * for the server's Retry-After (default 60s) instead of burning retries
 * against it. Key values never appear in stats or error labels.
 */

const DEFAULT_RETRY_AFTER_MS = 60_000;

export type FlockKeyStats = {
  /** Opaque label ("key#1") — never the key value. */
  label: string;
  successCount: number;
  failureCount: number;
  rateLimitCount: number;
  rateLimitedUntil: number | null;
  isRateLimited: boolean;
  isAvailable: boolean;
  lastFailureReason: string | null;
};

type KeyState = {
  key: string;
  successCount: number;
  failureCount: number;
  rateLimitCount: number;
  rateLimitedUntil: number | null;
  lastFailureReason: string | null;
};

export class FlockKeyPool {
  private keys: KeyState[] = [];
  private cursor = 0;

  constructor(keys: string[]) {
    const seen = new Set<string>();
    for (const raw of keys) {
      const key = raw.trim();
      if (!key || seen.has(key)) continue;
      seen.add(key);
      this.keys.push({
        key,
        successCount: 0,
        failureCount: 0,
        rateLimitCount: 0,
        rateLimitedUntil: null,
        lastFailureReason: null,
      });
    }
  }

  get size(): number {
    return this.keys.length;
  }

  private isLimited(state: KeyState, now: number): boolean {
    return state.rateLimitedUntil !== null && state.rateLimitedUntil > now;
  }

  /** Next available key (round-robin), or null when all are rate-limited. */
  nextAvailable(): string | null {
    const now = Date.now();
    for (let i = 0; i < this.keys.length; i++) {
      const idx = (this.cursor + i) % this.keys.length;
      if (!this.isLimited(this.keys[idx], now)) {
        this.cursor = (idx + 1) % this.keys.length;
        return this.keys[idx].key;
      }
    }
    return null;
  }

  recordSuccess(key: string): void {
    const s = this.keys.find((k) => k.key === key);
    if (!s) return;
    s.successCount++;
    s.rateLimitedUntil = null;
    s.lastFailureReason = null;
  }

  recordFailure(key: string, reason: string): void {
    const s = this.keys.find((k) => k.key === key);
    if (!s) return;
    s.failureCount++;
    s.lastFailureReason = reason;
  }

  recordRateLimit(key: string, retryAfterMs?: number): void {
    const s = this.keys.find((k) => k.key === key);
    if (!s) return;
    s.rateLimitCount++;
    s.lastFailureReason = "Rate limited (429)";
    s.rateLimitedUntil = Date.now() + (retryAfterMs ?? DEFAULT_RETRY_AFTER_MS);
  }

  /** Earliest ms-epoch at which a rate-limited key becomes usable, or null. */
  earliestRetryAt(): number | null {
    const times = this.keys
      .map((k) => k.rateLimitedUntil)
      .filter((t): t is number => t !== null);
    return times.length ? Math.min(...times) : null;
  }

  getStats(): FlockKeyStats[] {
    const now = Date.now();
    return this.keys.map((s, i) => {
      const limited = this.isLimited(s, now);
      return {
        label: "key#" + (i + 1),
        successCount: s.successCount,
        failureCount: s.failureCount,
        rateLimitCount: s.rateLimitCount,
        rateLimitedUntil: s.rateLimitedUntil,
        isRateLimited: limited,
        isAvailable: !limited,
        lastFailureReason: s.lastFailureReason,
      };
    });
  }
}

/** Split a comma-separated key string into deduplicated keys. */
export function splitKeys(raw: string | string[] | undefined): string[] {
  const list = Array.isArray(raw) ? raw : raw ? raw.split(",") : [];
  const seen = new Set<string>();
  const out: string[] = [];
  for (const k of list) {
    const key = k.trim();
    if (key && !seen.has(key)) {
      seen.add(key);
      out.push(key);
    }
  }
  return out;
}

/** Parse a Retry-After header (seconds or HTTP date) into ms. */
export function parseRetryAfterMs(value: string | null): number {
  if (!value) return DEFAULT_RETRY_AFTER_MS;
  const secs = Number(value.trim());
  if (Number.isFinite(secs) && secs >= 0) return secs * 1000;
  const dateMs = Date.parse(value);
  if (!Number.isNaN(dateMs)) return Math.max(0, dateMs - Date.now());
  return DEFAULT_RETRY_AFTER_MS;
}

File: client/python/flock_client/keypool.py
"""FlockKeyPool — multi-key rotation with 429/Retry-After tracking.

Python mirror of the TypeScript flock-client keypool: keys are deduplicated
on registration, rotation is round-robin over keys whose rate limit has
expired, and a 429 parks the offending key for the server's Retry-After
(default 60s). Key values never appear in stats or error labels.
"""

from __future__ import annotations

import time
from email.utils import parsedate_to_datetime

DEFAULT_RETRY_AFTER_MS = 60_000


def split_keys(raw: "str | list[str] | None") -> list[str]:
    items = raw if isinstance(raw, list) else (raw.split(",") if raw else [])
    seen: set[str] = set()
    out: list[str] = []
    for item in items:
        key = item.strip()
        if key and key not in seen:
            seen.add(key)
            out.append(key)
    return out


def parse_retry_after_ms(value: "str | None") -> int:
    if not value:
        return DEFAULT_RETRY_AFTER_MS
    value = value.strip()
    try:
        secs = float(value)
        if secs >= 0:
            return int(secs * 1000)
    except ValueError:
        pass
    try:
        dt = parsedate_to_datetime(value)
        return max(0, int(dt.timestamp() * 1000 - time.time() * 1000))
    except (ValueError, TypeError):
        return DEFAULT_RETRY_AFTER_MS


class FlockKeyPool:
    def __init__(self, keys: "list[str]"):
        seen: set[str] = set()
        self._keys: list[dict] = []
        for raw in keys:
            key = raw.strip()
            if not key or key in seen:
                continue
            seen.add(key)
            self._keys.append(
                {
                    "key": key,
                    "success_count": 0,
                    "failure_count": 0,
                    "rate_limit_count": 0,
                    "rate_limited_until": None,
                    "last_failure_reason": None,
                }
            )
        self._cursor = 0

    @property
    def size(self) -> int:
        return len(self._keys)

    @staticmethod
    def _limited(state: dict, now: float) -> bool:
        until = state["rate_limited_until"]
        return until is not None and until > now

    def next_available(self) -> "str | None":
        now = time.time()
        for i in range(len(self._keys)):
            idx = (self._cursor + i) % len(self._keys)
            if not self._limited(self._keys[idx], now):
                self._cursor = (idx + 1) % len(self._keys)
                return self._keys[idx]["key"]
        return None

    def _find(self, key: str) -> "dict | None":
        for s in self._keys:
            if s["key"] == key:
                return s
        return None

    def record_success(self, key: str) -> None:
        s = self._find(key)
        if s:
            s["success_count"] += 1
            s["rate_limited_until"] = None
            s["last_failure_reason"] = None

    def record_failure(self, key: str, reason: str) -> None:
        s = self._find(key)
        if s:
            s["failure_count"] += 1
            s["last_failure_reason"] = reason

    def record_rate_limit(self, key: str, retry_after_ms: "int | None" = None) -> None:
        s = self._find(key)
        if s:
            s["rate_limit_count"] += 1
            s["last_failure_reason"] = "Rate limited (429)"
            s["rate_limited_until"] = time.time() + (
                retry_after_ms if retry_after_ms is not None else DEFAULT_RETRY_AFTER_MS
            ) / 1000.0

    def earliest_retry_at(self) -> "float | None":
        times = [s["rate_limited_until"] for s in self._keys if s["rate_limited_until"] is not None]
        return min(times) if times else None

    def get_stats(self) -> list[dict]:
        now = time.time()
        return [
            {
                "label": f"key#{i + 1}",
                "success_count": s["success_count"],
                "failure_count": s["failure_count"],
                "rate_limit_count": s["rate_limit_count"],
                "rate_limited_until": s["rate_limited_until"],
                "is_rate_limited": self._limited(s, now),
                "is_available": not self._limited(s, now),
                "last_failure_reason": s["last_failure_reason"],
            }
            for i, s in enumerate(self._keys)
        ]
